### ISSUE TITLE
When using C-FIND, how to set the character encoding? 

### ISSUE THREAD

**Issue by suoc** [PantelisGeorgiadis/dcmjs-dimse#60]
When using C-FIND , how to set the character encoding? Garbled characters may appear when using Chinese or other language characters.
The character encoding set by the peer SCU is GB18030.For example, when the patient's name is '测试', garbled characters will appear when returned to the SCU end.

**Comment by PantelisGeorgiadis** [comment]
Hello @suoc,
Thank you for your question! Setting the character encoding is easy... all you have to do is to set the response's `SpecificCharacterSet` tag to the character set value you need, like the following:
```js
const pendingResponse = CFindResponse.fromRequest(request);
pendingResponse.setDataset(new Dataset({ SpecificCharacterSet: 'GB18030', PatientID: '12345', PatientName: 'JOHN^DOE' }));
pendingResponse.setStatus(Status.Success);
```
The hard part would be to provide the right bytes for the "encodable" VRs (SH, LO, ST, PN, LT, UC or UT) of the specified character encoding. Unfortunately, `dcmjs` doesn't provide such functionality, as far as I know and, therefore, you need to use an external library (e.g. `iconv`, `TextEncoder`, etc.) and figure a way to inject the character bytes in the tag values.

**Comment by PantelisGeorgiadis** [comment]
Hello @suoc,
Did you manage to properly encode your datasets so you can return strings represented in specific character sets?

**Comment by suoc** [comment]
OK, I'll try it. I haven't looked at this question recently. I'll give you feedback when I get the result.

**Comment by suoc** [comment]
After testing, the character encoding can only be set to SpecificCharacterSet: 'ISO_IR 192', which requires the same support from the device. By viewing the source code and test cases of dcmjs, it can be seen that dcmjs does not perform character encoding according to the configuration of SpecificCharacterSet.

**Comment by PantelisGeorgiadis** [comment]
Right, `dcmjs` does not perform character encoding according to the configuration of `SpecificCharacterSet`. That's why I mentioned that "you need to use an external library (e.g. iconv, TextEncoder, etc.) and figure a way to inject the character bytes in the tag values". I would be very interested to see if this is feasible!

**Comment by suoc** [comment]
I have an idea to implement character encoding and decoding through global configuration, instead of integrating character encoding and decoding libraries everywhere in the source code.
The approximate code is as follows:
```javascript
dcmjs.decoder = function(buffer, specificCharacterSet) {
    // .............................
}

dcmjs.encoder = function(tag_value, specificCharacterSet) {
    // .............................
}
```
Is this approach feasible?

**Comment by PantelisGeorgiadis** [comment]
That sounds like a great idea @suoc! Your functions should also be aware of the element that are encoding/decoding since not all VRs are encodable (only SH, LO, ST, PN, LT, UC and UT). @pieper are you aware of any similar effort?

**Comment by pieper** [comment]
> are you aware of any similar effort?

No, but I agree it would be great to see support for this (along with tests of course : )

**Comment by PantelisGeorgiadis** [comment]
Hello @suoc! Any developments on encoding/decoding your datasets prior pushing them?

**Comment by suoc** [comment]
I have attempted to integrate iconv-lite into dcmjs to solve the character encoding and decoding problem. The test results are positive, but the integration is not elegant enough. I still need to take the time to sort out the code coupling problem.
```javascript
export class IconvDecoder {
    constructor(encoding) {
        this.encoding = encoding ? encoding : "GB18030";
    }

    decode(buf) {
        let iconv = require("iconv-lite");
        return iconv.decode(Buffer.from(buf), this.encoding);
    }
}
export class IconvEncoder {
    constructor(encoding) {
        this.encoding = encoding ? encoding : "GB18030";
    }

    encode(str) {
        let iconv = require("iconv-lite");
        return iconv.encode(str, this.encoding);
    }
}

export function converToDcmCodingName(js_coding_name) {
    if (isEqualToAny(js_coding_name, 'utf8', 'utf-8')) {
        return 'ISO_IR 192';
    }
    js_coding_name = js_coding_name.toLowerCase();
    let dcm_names = Object.keys(encodingMapping);
    for (let i = 0; i < dcm_names.length; i++) {
        let dcm_coding_name = dcm_names[i];
        if (encodingMapping[dcm_coding_name] === js_coding_name) {
            dcm_coding_name = dcm_coding_name ? dcm_coding_name : 'ISO-8859-1';
            return dcm_coding_name.replace('iso-ir-', 'ISO_IR ').toUpperCase();
        }
    }
}

export function converToJsCodingName(dcm_coding_name) {
    dcm_coding_name = dcm_coding_name.replace(/[_ ]/g, "-").toLowerCase();
    return encodingMapping[dcm_coding_name];
}
```

**Comment by PantelisGeorgiadis** [comment]
Hello @suoc! 
Great effort! I believe that you need to create a pull request in the dcmjs repository to get feedback and ideas from the people that actually created and maintaining the library. I am pretty confident that after implementing this feedback, the solution is going to be very elegant!

**Comment by PantelisGeorgiadis** [comment]
Hello @suoc! I'm closing this issue now. It would be great if you can work with the core `dcmjs` development team and bring such a great feature to the library!
